compiler/optimizer/VPHandlers.cpp

```diff
@@ -1342,19 +1342,26 @@ bool simplifyJ9ClassFlags(OMR::ValuePropagation *vp, TR::Node *node, bool isLong
 
    if (symRef == vp->comp()->getSymRefTab()->findClassAndDepthFlagsSymbolRef())
       {
+      TR::Node *baseNode = NULL;
       if (node->getFirstChild()->getOpCode().hasSymbolReference() &&
             (node->getFirstChild()->getSymbolReference() == vp->comp()->getSymRefTab()->findVftSymbolRef()))
          {
-         TR::VPConstraint *baseObject = vp->getConstraint(node->getFirstChild()->getFirstChild(), isGlobal);
-         if (baseObject && baseObject->getClassType() &&
-               baseObject->getClassType()->asFixedClass())
-            {
-            cdfValue = vp->comp()->fej9()->getClassDepthAndFlagsValue(baseObject->getClassType()->getClass());
-            // if the type is a java.lang.Object, make the control branch to the call
-            //
-            if (baseObject->getClassType()->asFixedClass()->isJavaLangObject(vp))
-               cdfValue = TR::Compiler->cls.flagValueForFinalizerCheck(vp->comp());
-            }
+         baseNode = node->getFirstChild()->getFirstChild();
+         }
+      else if(node->getFirstChild()->getOpCode().isLoadAddr() && node->getFirstChild()->getSymbolReference()->getSymbol()->isClassObject())
```

Comment:
I think this logic probably shouldn't have to inspect the structure of the child at all, not even with the original `if` just above. We should be able to just ask for the constraint of `node->getFirstChild()`, e.g.
```
TR::VPConstraint *base = vp->getConstraint(node->getFirstChild(), isGlobal);
if (base != NULL && base->isJ9ClassObject() == TR_yes && base->isFixedClass())
   ...
```

Reply:
Yes, that should work as well. Thanks a lot for the suggestions. Let me try it out.

Reply:
@jdmpapin  Made changes in https://github.com/eclipse/omr/compare/f94a66015781949772cf21347d57a624b2af2bee..abdec2fc9577133c9b9f9c052b17978b14f2e9e9 and I can confirm with the test in which it was failing, and it works as expected.